### ISSUE TITLE
Investigate possible storage incompatibilities with MinIO

### ISSUE THREAD

**Issue by gabek** [owncast/owncast#242]
[This tutorial](https://azhuge233.com/owncast-%E5%BF%AB%E9%80%9F%E6%90%AD%E5%BB%BA%E4%B8%AA%E4%BA%BA%E7%9B%B4%E6%92%AD%E5%B9%B3%E5%8F%B0/) when translated had this noted:

```s3:
    endpoint: http://192.168.16.16:9000
    accessKey: azhuge233
    secret: veryStrongSecretKey
    region: us-east-1
    bucket: owncast
```

>Three levels of streaming quality and S3 storage have been added. MinIO can be used for S3 storage. For details, see Setup private object storage using MinIO
>
>However, if you use MinIO to open S3 storage according to the above configuration, the stream will not be loaded. I don’t know why

I've never used [MinIO](https://min.io/), and I'm not sure how popular it is to host your own S3 storage, but it might be worth looking into to see what issues there might be.

**Comment by jeyemwey** [comment]
### CAUTION: This is not a good tutorial for using Minio securely with Owncast!

It took some fiddle, but this is how I got Owncast+MinIO working on my local machine:

```bash
$ docker run -it --rm -p 8080:8080 -p 1935:1935 -v $PWD/config.yaml:/app/config.yaml master_build_of_owncast /app/owncast`
$ docker run -d -p 9000:9000 -v $PWD/minio:/data -e "MINIO_ACCESS_KEY=ABCDEFGABCDEFG" -e "MINIO_SECRET_KEY=abcdefgabcdefg" -e "MINIO_REGION_NAME=de" minio/minio server /data`
```

Docker didn't let Owncast connect to Minio via the `127.0.0.1` address, but it worked with the `172.16.0.0/12` address. For the serving endpoint, the link-local address is fine.

```yaml
s3:
   enabled: true
   endpoint:  http://172.17.0.2:9000
   servingEndpoint: http://127.0.0.1:9000/owncast
   accessKey: ABCDEFGABCDEFG
   secret: abcdefgabcdefg
   region: de
   bucket: owncast
```

Before starting Owncast, the bucket had to be created in the Minio Web Interface and granted "Read/Write access". I have probably removed all possible security from the bucket, so I will not exacly tell you how this is possible.

After that, the server started, accepted and played video just fine. Not sure what the problem is here.

**Comment by gabek** [comment]
Thank you so much for testing this!

So it sounds like there is no inherent incompatibility with using MinIO.  MinIO seems to have its own permissions quirks, but I'm not sure that has anything to do with Owncast and using it for storage.

I think we can say confidently after your test that it works, but if you think there's anything more we should do, or any additional documentation we could do to make things better we can do that.  But it sounds like it works.

**Comment by gabek** [comment]
Going to close this since it sounds like there's no inherent incompatibility between MinIO and Owncast.  Feel free to open if there's anything anybody runs into that we need to address.

**Comment by stsecurity** [comment]
Hello, I'm using v0.0.10 and after setting up everything to connect to my minio address, including 'Serving Endpoint', owncast is still trying to connect to 'https://bucketname.myminiodomain' instead of 'https://myminiodomain/bucketname' which is what I set in my admin panel.
It feels like the 'Serving Endpoint' option does nothing, is there anything I'm missing or is it a bug? I looked at the docs but it seems v0.0.10 removed the config file so I cannot modify the config directly.

**Comment by gabek** [comment]
Serving endpoint was designed for S3 providers that offer a CDN where you need to pull the files from an alternative hostname. Admittedly I'm not sure what would happen if you tried to add a path to that as well, like you're doing. I'm guessing it breaks things since that's not what it was built for.

You're running into the "Host based vs. Path based" S3 scenario. Owncast currently only supports host based, not path based, S3 configurations. However in 0.0.11 you'll be able to switch between the two.  

I've never used Minio, and I can't offer much support. But I did a search for "minio host-based" and got this page https://docs.min.io/docs/minio-server-configuration-guide.html where it says 

> By default, MinIO supports path-style requests that are of the format http://mydomain.com/bucket/object. MINIO_DOMAIN environment variable is used to enable virtual-host-style requests. If the request Host header matches with (.+).mydomain.com then the matched pattern $1 is used as bucket and the path is used as object. More information on path-style and virtual-host-style here

So it sounds like a Minio configuration change might fix your issues.

<blockquote><div><strong><a href="https://docs.min.io/docs/minio-server-configuration-guide.html">MinIO | Learn how to configure your MinIO server</a></strong></div><div>This guide includes code, examples and tips on how to get the most out of your MinIO server</div></blockquote>
<blockquote><div><strong><a href="http://mydomain.com/bucket/object">404 Error | MyDomain.com</a></strong></div><div>Small business web hosting offering additional business services such as: domain name registrations, email accounts, web services, online community resources and various small business solutions.</div></blockquote>

**Comment by stsecurity** [comment]
> Serving endpoint was designed for S3 providers that offer a CDN where you need to pull the files from an alternative hostname. Admittedly I'm not sure what would happen if you tried to add a path to that as well, like you're doing. I'm guessing it breaks things since that's not what it was built for.
> 
> You're running into the "Host based vs. Path based" S3 scenario. Owncast currently only supports host based, not path based, S3 configurations. However in 0.0.11 you'll be able to switch between the two.
> 
> I've never used Minio, and I can't offer much support. But I did a search for "minio host-based" and got this page https://docs.min.io/docs/minio-server-configuration-guide.html where it says
> 
> > By default, MinIO supports path-style requests that are of the format http://mydomain.com/bucket/object. MINIO_DOMAIN environment variable is used to enable virtual-host-style requests. If the request Host header matches with (.+).mydomain.com then the matched pattern $1 is used as bucket and the path is used as object. More information on path-style and virtual-host-style here
> 
> So it sounds like a Minio configuration change might fix your issues.
> 
> > **[MinIO | Learn how to configure your MinIO server](https://docs.min.io/docs/minio-server-configuration-guide.html)**This guide includes code, examples and tips on how to get the most out of your MinIO server
> 
> > **[404 Error | MyDomain.com](http://mydomain.com/bucket/object)**Small business web hosting offering additional business services such as: domain name registrations, email accounts, web services, online community resources and various small business solutions.

Thank you very much for your reply, I was trying to set serving endpoint because it was mentioned in [owncast docs](https://owncast.online/docs/storage/minio/), but I guess it was written for an earlier version.
I do understand that MinIO support host-based method, for me I can't easily switch to a host-based method because I use my MinIO instance for some other services I run and I want to keep the paths consistent. Also I used to use cloudflare CDN for my MinIO instance, they don't support second level subdomain for their free plans, which could be another reason for people wanting to use the path style configuration, if their MinIO instance is hosted on a subdomain.
Anyways knowing that the setting will be available in v0.0.11 is enough for me, thanks again for your reply. Have a nice day.

<blockquote><div><strong><a href="https://docs.min.io/docs/minio-server-configuration-guide.html">MinIO | Learn how to configure your MinIO server</a></strong></div><div>This guide includes code, examples and tips on how to get the most out of your MinIO server</div></blockquote>
<blockquote><div><strong><a href="http://mydomain.com/bucket/object">404 Error | MyDomain.com</a></strong></div><div>Small business web hosting offering additional business services such as: domain name registrations, email accounts, web services, online community resources and various small business solutions.</div></blockquote>
